include/boost/json/basic_parser_impl.hpp

```diff
@@ -1398,19 +1428,43 @@ parse_escaped(
                     = BOOST_CURRENT_LOCATION;
                 return fail(cs.begin(), error::expected_hex_digit, &loc);
             }
-            unsigned const u2 =
+            unsigned u2 =
                 (d1 << 12) + (d2 << 8) +
                 (d3 << 4) + d4;
-            // valid trailing surrogates are [DC00, DFFF]
+            // Check if the second code point is a valid trailing surrogate.
+            // Valid trailing surrogates are [DC00, DFFF]
             if(BOOST_JSON_UNLIKELY(
-                u2 < 0xdc00 || u2 > 0xdfff))
+                    u2 < 0xdc00 || u2 > 0xdfff))
             {
-                BOOST_STATIC_CONSTEXPR source_location loc
-                    = BOOST_CURRENT_LOCATION;
-                return fail(cs.begin(), error::illegal_trailing_surrogate,
-                    &loc);
+                // If not valid and the parser does not allow it, return an error.
+                if(!allow_bad_utf16)
+                {
+                    BOOST_STATIC_CONSTEXPR source_location loc
+                        = BOOST_CURRENT_LOCATION;
+                    return fail(cs.begin(), error::illegal_trailing_surrogate,
+                        &loc);
+                }
+                // Append the replacement character for the
+                // first leading surrogate.
+                cs += 4;
+                temp.append_utf8(urc);
+                // Check if the second code point is also not
+                // a valid leading surrogate.
+                if (u2 < 0xd800 || u2 > 0xdbff)
+                {
+                    temp.append_utf8(u2);
+                    break;
+                }
+                // If it is a valid leading surrogate
+                else
+                {
+                    u1 = u2;
+                    goto do_str9;
```

Comment:
If you go there you potentially violate the assumption that `cs.remain() > 10`, and so you can overrun the input buffer.

Reply:
Since the value of the second code point **u2** is not stored in temp, how can we preserve it if we don't use goto and break instead?

Reply:
You can assign to `u1_` and go to `do_sur1`.